modules/standard/GPU.chpl

```diff
@@ -626,4 +626,191 @@ module GPU
   */
   inline proc gpuMaxLocReduce(const ref A: [] ?t) do return doGpuReduce("maxloc", A);
 
+
+
+  // ============================
+  // GPU Scans
+  // ============================
+
+  // The following functions are used to implement GPU scans. They are
+  // intended to be called from a GPU locale.
+  import BitOps;
+  import Math;
+
+
+  proc gpuScan(ref gpuArr: [] uint){
+    if gpuArr.size==0 then return;
+    // Use a simple algorithm for small arrays
+    // TODO check the actual thread block size rather than the2*default
+    if gpuArr.size <= 1024 {
+      // The algorithms only works for arrays that are size of a power of two.
+      // In case it's not a power of two we pad it out with 0s
+      const size = roundToPowerof2(gpuArr.size);
+      var arr : [0..<size] uint;
+      arr[0..<gpuArr.size] = gpuArr;
+
+      // Hillis Steele Scan is better if we can scan in
+      // a single thread block
+      // TODO check the actual thread block size rather than the default
+      if gpuArr.size <= 512 then
+        hillisSteeleScan(arr);
+      else
+        blellochScan(arr);
+
+      // Copy back
+      gpuArr=arr[0..<gpuArr.size];
+    } else {
+      // We use a parallel scan algorithm for large arrays
+      parallelArrScan(gpuArr);
+    }
+  }
+
+  private proc parallelArrScan(ref gpuArr: [] uint){
+    // Divide up the array into chunks of a reasonable size
+    // For our default, we choose our default block size which is 512
+    const scanChunkSize = 512;
+    const numScanChunks = Math.divCeil(gpuArr.size, scanChunkSize);
+
+    if numScanChunks == 1 {
+      hillisSteeleScan(gpuArr);
+      return;
+    }
+
+    // Allocate an accumulator array
+    var gpuScanArr : [0..<numScanChunks] uint;
+    const low = gpuArr.domain.low; // https://github.com/chapel-lang/chapel/issues/22433
+
+    const ceil = gpuArr.domain.high;
+    // In parallel: For each chunk we do an in lane serial scan
+    @assertOnGpu
+    foreach chunk in 0..<numScanChunks {
+      const start = low+chunk*scanChunkSize;
+      var end = start+scanChunkSize-1;
```

Comment:
You could do this as a one-liner to avoid making `end` a `var`:

const end = min(ceil, start+scanChunkSize-1)